File: AlgorithmsDemo/SelectionSort/selection_sort.py
# Topic: Selection sort algorithm implementation and demostration
#
# To run the Python file:
# $ python --version
# $ python .\selection_sort.py
# 
# To import the function from other Python source code:
# from selection_sort.py import selection_sort
#
# Author: Xuhua Huang
# Last updated: Nov 13, 2021
# Created on: Nov 13, 2021

from typing import Any


def selection_sort(_list: list[Any]):
    # Time complexity: O(n^2)
    # suppose that the first element is the smallest
    smallest_num_index: int = 0
    # loop through the list
    # for each iteration, find the index of the smaller number
    for i in range(len(_list)):
        smallest_num_index = i;
        for j in range(i+1, len(_list), 1):
            if _list[j] < _list[smallest_num_index]:
                smallest_num_index = j
        # if there is indeed a smaller number
        # use the updated index to swap the elements
        if smallest_num_index != i:
            _list[i], _list[smallest_num_index] = _list[smallest_num_index], _list[i]
    # add empty return statement for consistency
    return



def main():
    to_sort_int: list[int] = [4, 2, 7, 1, 3]
    print("Before calling selection_sort(): ", to_sort_int)
    selection_sort(to_sort_int)
    print("After calling selction_sort(): ", to_sort_int)

    to_sort_char: list[str] = ['a', 's', 'd', 'f', 'b', 'c', 'e']
    print("Before calling selection_sort(): ", to_sort_char)
    selection_sort(to_sort_char)
    print("After calling selection_sort(): ", to_sort_char)


if __name__ == '__main__':
    main()

File: AlgorithmsDemo/SelectionSort/selection_sort.c
/**
 * @file selection_sort.c
 * @author Xuhua Huang (xuhuahuang0412@gmail.com)
 * @brief Selection sort algorithm implementation and demonstration in C.
 * Time complexity: o(N^2)
 * Caution: facing random poitner glitches. Waiting to implement in C++.
 * 
 * To compile and run on Windows with Makefile:
 * $ mingw32-make
 * $ ./selectionsort
 * 
 * @version 0.1
 * @date 2021-11-14
 * 
 * @copyright Copyright (c) 2021
 * 
 */

#include <stdio.h>
#include <stdlib.h>

void selection_sort(int* arr) {
    int smallest_num_index = 0; /* Suppose the first element of the array is the smallest. */
    
    for (int i = 0; i < sizeof(arr); i++) {
        smallest_num_index = i; /* Update the index during each iteration. */
        /* Loop through the rest of the array. */
        for (int j = i; j < sizeof(arr); j++) {
            if (arr[j] < arr[smallest_num_index]) {
                smallest_num_index = j;
            }
        }
        /**
         * @brief If a smaller number is found, i.e.
         * variable `smallest_num_index` has been updated
         */
        if (smallest_num_index != i) {
            /* Swap the elements. */
            int temp = arr[i];
            arr[i] = arr[smallest_num_index];
            arr[smallest_num_index] = temp;
        }
    }

    return;
}

void print_array(int* arr) {
    printf("{ ");
    for (int i = 0; i < sizeof(arr); i++) {
        printf("%d ", arr[i]);
    }
    printf("}\n");
}

int main(void) {
    int to_sort[] = { 4, 2, 7, 1, 3 };
    print_array(to_sort);

    selection_sort(to_sort);
    print_array(to_sort);

    system("pause");
    return 0;
}